source/server/listener_impl.cc

```diff
@@ -353,20 +350,22 @@ ListenerImpl::ListenerImpl(const envoy::config::listener::v3::Listener& config,
       quic_stat_names_(parent_.quicStatNames()),
       missing_listener_config_stats_({ALL_MISSING_LISTENER_CONFIG_STATS(
           POOL_COUNTER(listener_factory_context_->listenerScope()))}) {
-
-  if ((address_->type() == Network::Address::Type::Ip &&
-       config.address().socket_address().ipv4_compat()) &&
-      (address_->ip()->version() != Network::Address::IpVersion::v6 ||
-       (!address_->ip()->isAnyAddress() &&
-        address_->ip()->ipv6()->v4CompatibleAddress() == nullptr))) {
-    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.strict_check_on_ipv4_compat")) {
-      throw EnvoyException(fmt::format(
-          "Only IPv6 address '::' or valid IPv4-mapped IPv6 address can set ipv4_compat: {}",
-          address_->asStringView()));
-    } else {
-      ENVOY_LOG(warn, "An invalid IPv4-mapped IPv6 address is used when ipv4_compat is set: {}",
-                address_->asStringView());
+  // All the addresses should be same socket type, so get the first address's socket type is enough.
```

Comment:
Maybe I miss the context, is it designed like this or will it be optimized in the future?

Reply:
yea, this is defined in the API, you can reference here

https://github.com/envoyproxy/envoy/blob/e5ef611dd9a0e840681cd6e99c6cffc97ecb47b5/api/envoy/config/listener/v3/listener.proto#L124-L128